drivers: pinctrl: rp2040: oe-override option

yiding opened the pull request:
This change adds the device tree property for specifying oe-override (output-enable override behavior), as well as defines for possible values of the property.

RP2040 GPIOs can be configured to automatically invert the output-enable signal from the selected peripheral function. This is useful for tasks like writing efficient PIO code, such as in the i2c example in the rp2040 datasheet.

I've manually tested the functionality with value set to INVERT. Not sure if there's any automated tests I should update too.

decsny commented on dts/bindings/pinctrl/raspberrypi,pico-pinctrl.yaml:
```diff
@@ -121,3 +121,17 @@ child-binding:
+        default: 0
+        description: |
+          Override output-enable for a pin.
+          0 - drive output enable from selected peripheral signal (default)
```
justify default https://docs.zephyrproject.org/latest/build/dts/bindings-upstream.html#rules-for-default-values

aasinclair commented on dts/bindings/pinctrl/raspberrypi,pico-pinctrl.yaml:
```diff
@@ -121,3 +121,17 @@ child-binding:
+          Override output-enable for a pin.
+          0 - drive output enable from selected peripheral signal (default)
+          1 - drive output enable from inverse of selected peripheral signal.
+          2 - disable output
```
From the datasheet 2/3 look like active low / active high, is that correct?

yiding replied:
I'm not sure what you mean by active low or high. My reading is that setting 2 or 3 means "ignore whether peripheral is sending, behave as if you are getting a low (in the case of 2) or high (in the case of 3) signal".

In the case of OEOVER (the subject of this PR), this means either disabling the output, or enabling the output. 

Source: Rp2040 datasheet, Page 247 (IO BANK0: GPIO0_CTRL, ...)

aasinclair replied:
Sorry, I was reading the wrong field in the datasheet.

aasinclair commented on include/zephyr/dt-bindings/pinctrl/rpi-pico-rp2040-pinctrl.h:
```diff
@@ -25,6 +25,11 @@
 #define RP2_PIN_NUM_POS 4
 #define RP2_PIN_NUM_MASK 0x1f
 
+#define RP2_GPIO_OVERRIDE_NORMAL 0
```
These defines don't appear to be used.

yiding replied:
Normal, Invert, High, Low are the 4 possible values (they are named this way in the rp2 SDK). The intent is that users setting this field would use the options instead of the numeric constants. I could document this in the yaml file instead of using the numbers.

aasinclair reviewed (dismissed):
Looks good, thanks for adding this.

gmarull commented on dts/bindings/pinctrl/raspberrypi,pico-pinctrl.yaml:
```diff
@@ -121,3 +121,22 @@ child-binding:
         description: |
           The slew rate of a pin. 0 corresponds to slow, and 1 corresponds to fast.
           The default value is 0 (slow), as this is the power on reset value.
+      oe-override:
```
please, use vendor prefix for custom properties.

yiding replied:
Updated to use rp2-, which seems consistent with the prefix used for defines in the header file.

gmarull commented on dts/bindings/pinctrl/raspberrypi,pico-pinctrl.yaml:
```diff
@@ -121,3 +121,22 @@ child-binding:
         description: |
           The slew rate of a pin. 0 corresponds to slow, and 1 corresponds to fast.
           The default value is 0 (slow), as this is the power on reset value.
+      rp2-oe-override:
```
vendor prefix should be `raspberry,`

yiding committed:
drivers: pinctrl: rp2040: oe-override option

This change adds the device tree property for specifying oe-override
(output-enable override behavior), as well as defines for possible values
of the property.

RP2040 GPIOs can be configured to automatically invert the output-enable
signal from the selected peripheral function. This is useful for tasks like
writing efficient PIO code, such as in the i2c example in the rp2040
datasheet.


Signed-off-by: Yiding Jia <yiding.jia@gmail.com>
